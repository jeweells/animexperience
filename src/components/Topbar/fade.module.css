.enter{
    opacity: 0;
}
.exit{
    opacity: 1;
}
.enterActive{
    opacity: 1;
}
.exitActive{
    opacity: 0;
}
.enterActive,
.exitActive{
    transition: opacity 200ms;
}