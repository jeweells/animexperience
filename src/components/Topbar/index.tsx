import React, { Fragment, useLayoutEffect, useState } from 'react'
import { CSSTransition, SwitchTransition } from 'react-transition-group'
import styled from 'styled-components'
import { useAppSelector } from '../../../redux/store'
import fade from './fade.module.less'
const Wrapper = styled.div`
    background: #25282e;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    z-index: 100000;
    -webkit-user-select: none;
    -webkit-app-region: drag;
`

const Container = styled.div`
    padding: 16px 24px;
    font-size: 1rem;
    font-weight: 700;
`

export const useTopBarHeight = () => 56

export type TopbarProps = {}

export const Topbar: React.FC<TopbarProps> = React.memo(({ children }) => {
    const watching = useAppSelector((d) => d.watch.watching)
    const [isFullscreen, setIsFullScreen] = useState(false)
    useLayoutEffect(() => {
        const handle = () => {
            setIsFullScreen(!!document.fullscreenElement)
        }
        document.addEventListener('fullscreenchange', handle)
        return () => {
            document.removeEventListener('fullscreenchange', handle)
        }
    }, [])
    const height = useTopBarHeight()
    const title = watching
        ? `${watching.name} - Episodio ${watching.episode}`
        : 'Animexperience'
    return (
        <Fragment>
            <Wrapper style={{ height, display: isFullscreen ? 'none' : 'block' }}>
                <Container>
                    <SwitchTransition>
                        <CSSTransition
                            key={title}
                            addEndListener={(node, done) =>
                                node.addEventListener('transitionend', done, false)
                            }
                            classNames={{ ...fade }}
                        >
                            <div>{title}</div>
                        </CSSTransition>
                    </SwitchTransition>
                </Container>
            </Wrapper>
            <div style={{ paddingTop: height }}>{children}</div>
        </Fragment>
    )
})

Topbar.displayName = 'Topbar'

export default Topbar
