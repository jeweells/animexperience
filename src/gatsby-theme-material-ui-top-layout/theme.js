import { createTheme } from '@material-ui/core/styles'

export const mainColor = 'rgb(105, 190, 213)'

const theme = createTheme({
    typography: {
        fontFamily: ['Quicksand', 'sans-serif'].join(','),
    },
    palette: {
        primary: {
            main: '#292d33',
            contrastText: '#e9ebf0',
        },
        secondary: {
            main: '#3c3f43',
        },
        text: {
            primary: '#e9ebf0',
            secondary: '#e9ebf0',
            disabled: '#ffffff82',
        },
    },
    overrides: {
        MuiButtonBase: {
            root: {
                borderRadius: undefined,
            },
        },
        MuiButton: {
            root: {
                color: '#e9ebf0',
                backgroundColor: '#292d33',
                marginBottom: 0,
                fontWeight: 'normal',
                textAlign: 'center',
                verticalAlign: 'middle',
                outline: '0 !important',
                whiteSpace: 'nowrap',
                border: 'none',
                userSelect: 'none',
                fontSize: 14,
                lineHeight: '1.42857143',
                borderRadius: 6,
                height: 36,
                '&:hover': {
                    backgroundColor: '#3c3f43',
                },
            },

            text: {
                color: '#e9ebf0',
                textTransform: 'none',
                padding: '8px 12px',
            },
        },
        MuiIconButton: {
            root: {
                flexShrink: 0,
                height: 36,
                minWidth: 36,
                color: '#e9ebf0',
                padding: '8px 12px',
                backgroundColor: '#292d33',
                marginBottom: 0,
                fontWeight: 'normal',
                textAlign: 'center',
                verticalAlign: 'middle',
                outline: '0 !important',
                whiteSpace: 'nowrap',
                border: 'none',
                userSelect: 'none',
                fontSize: 14,
                lineHeight: '1.42857143',
                borderRadius: 6,
                '&:hover': {
                    backgroundColor: '#3c3f43',
                },
            },
        },
        MuiCssBaseline: {
            '@global': {
                html: {
                    WebkitFontSmoothing: 'auto',
                },
                body: {
                    color: undefined,
                    margin: undefined,
                    fontSize: undefined,
                    fontFamily: undefined,
                    fontWeight: undefined,
                    backgroundColor: undefined,
                    lineHeight: undefined,
                },
            },
        },
        MuiSkeleton: {
            root: {
                backgroundColor: 'rgba(255, 255, 255, 0.11)',
            },
            pulse: {
                animation: 'MuiSkeleton-keyframes-pulse 3s ease-in 0.5s infinite',
            },
        },
    },
})

export default theme
