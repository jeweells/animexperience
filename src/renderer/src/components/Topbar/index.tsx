import {
  createContext,
  Fragment,
  PropsWithChildren,
  useContext,
  useRef,
  FC,
  memo,
  RefObject
} from 'react'
import { CSSTransition, SwitchTransition } from 'react-transition-group'
import { styled } from '@mui/system'
import { useAppSelector } from '~/redux/utils'
import CloseButton from '../../atoms/CloseButton'
import SearchBar from '../SearchBar'
import fade from './fade.module.css'
import ShareAnimeEpisodeButton from '../ShareAnimeEpisodeButton'
import { useIsFullscreen } from '~/src/hooks/useIsFullscreen'

const ipcRenderer = window.electron.ipcRenderer

const Wrapper = styled('div')`
  background: rgb(21, 24, 31);
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 100000;
  -webkit-user-select: none;
  -webkit-app-region: drag;
  display: flex;
  align-items: center;
  transition: height 300ms ease-in-out;
  will-change: height, opacity;
  contain: paint;
`

const Container = styled('div')`
  padding: 0 0 0 24px;
  font-size: 1rem;
  font-weight: 700;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 100%;
  flex: 1;
`
const Buttons = styled('div')`
  display: flex;
  align-items: center;
  gap: 8px;
  height: 100%;
`

const Content = styled('div')`
  overflow-x: hidden;
  position: relative;
`

export const useTopBarHeight = () => {
  const isFullscreen = useIsFullscreen()
  if (isFullscreen) return 0
  return 56
}

export type TopbarProps = PropsWithChildren

export const ContentContext = createContext<RefObject<HTMLDivElement> | null>(null)
export const useContentRef = () => useContext(ContentContext)

const closeApp = () => {
  return ipcRenderer.invoke('closeApp')
}

export const Topbar: FC<TopbarProps> = memo(({ children }) => {
  const watching = useAppSelector((d) => d.watch.watching)
  const height = useTopBarHeight()
  const title = watching ? `${watching.name} - Episodio ${watching.episode}` : 'ANIMEXPERIENCE'
  const contentRef = useRef<HTMLDivElement>(null)
  return (
    <Fragment>
      <Wrapper
        style={{
          height
        }}
      >
        <SwitchTransition>
          <CSSTransition
            key={title}
            addEndListener={(node, done) => node.addEventListener('transitionend', done, false)}
            classNames={{ ...fade }}
          >
            <Container>
              <div>{title}</div>
              {!watching && <SearchBar style={{ marginRight: 8 }} />}
              {watching && (
                <ShareAnimeEpisodeButton
                  style={{
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    WebkitAppRegion: 'no-drag',
                    WebkitUserSelect: 'all'
                  }}
                />
              )}
            </Container>
          </CSSTransition>
        </SwitchTransition>
        <Buttons>
          <CloseButton
            style={{
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              WebkitAppRegion: 'no-drag',
              WebkitUserSelect: 'all',
              borderRadius: 0,
              height: '100%',
              width: 48,

              background: 'rgb(0,0,0,0)'
            }}
            onClick={closeApp}
          />
        </Buttons>
      </Wrapper>
      <Content
        ref={contentRef}
        style={{
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          overflowY: 'overlay',
          marginTop: height,
          height: `calc(100vh - ${height}px)`
        }}
      >
        <ContentContext.Provider value={contentRef}>{children}</ContentContext.Provider>
      </Content>
    </Fragment>
  )
})

Topbar.displayName = 'Topbar'

export default Topbar
