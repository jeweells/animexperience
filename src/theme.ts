import { createTheme } from '@mui/material/styles'

const theme = createTheme({
    breakpoints: {
        values: {
            xs: 0,
            sm: 500,
            md: 800,
            lg: 1000,
            xl: 1200,
        },
    },
    typography: {
        fontFamily: ['Quicksand', 'sans-serif'].join(','),
    },
    palette: {
        mode: 'dark',
        primary: {
            main: '#292d33',
            contrastText: '#e9ebf0',
        },
        secondary: {
            main: '#3c3f43',
        },
        text: {
            primary: '#e9ebf0',
            secondary: '#e9ebf0',
            disabled: '#ffffff82',
        },
    },
    components: {
        MuiSvgIcon: {
            styleOverrides: {
                root: {
                    fontSize: 'inherit',
                },
            },
        },
        MuiButtonBase: {
            styleOverrides: {
                root: {
                    borderRadius: undefined,
                    textTransform: 'unset',
                },
            },
        },
        MuiButton: {
            styleOverrides: {
                root: {
                    color: '#e9ebf0',
                    backgroundColor: '#292d33',
                    marginBottom: 0,
                    fontWeight: 'normal',
                    textAlign: 'center',
                    verticalAlign: 'middle',
                    outline: '0 !important',
                    whiteSpace: 'nowrap',
                    border: 'none',
                    userSelect: 'none',
                    fontSize: 14,
                    lineHeight: '1.42857143',
                    textTransform: 'unset',
                    borderRadius: 6,
                    height: 36,
                    '&:hover': {
                        outline: 'none',
                        border: 'none',
                        backgroundColor: '#3c3f43',
                    },
                },
                text: {
                    color: '#e9ebf0',
                    textTransform: 'none',
                    padding: '8px 12px',
                },
            },
        },
        MuiIconButton: {
            styleOverrides: {
                root: {
                    flexShrink: 0,
                    height: 36,
                    minWidth: 36,
                    color: '#e9ebf0',
                    padding: '8px 6px',
                    backgroundColor: '#292d33',
                    marginBottom: 0,
                    fontWeight: 'normal',
                    textAlign: 'center',
                    verticalAlign: 'middle',
                    outline: '0 !important',
                    whiteSpace: 'nowrap',
                    border: 'none',
                    userSelect: 'none',
                    fontSize: '1.2rem',
                    lineHeight: '1.42857143',
                    borderRadius: 6,
                    '&:hover': {
                        backgroundColor: '#3c3f43',
                    },
                    '&.Mui-disabled': {
                        color: 'rgb(255 255 255 / 25%)',
                    },
                },
            },
        },
        MuiSkeleton: {
            defaultProps: {
                animation: 'pulse',
            },
            styleOverrides: {
                root: {
                    backgroundColor: 'rgba(255, 255, 255, 0.11)',
                },
                pulse: {
                    animationDuration: '1.5s 0.5s',
                },
            },
        },
        MuiCssBaseline: {
            styleOverrides: `
                * {
                    backface-visibility: hidden;
                    margin: 0;
                    padding: 0;
                    box-sizing: border-box;
                }
                html {
                    -webkit-font-smoothing: auto;
                }
                h6 {
                    margin: 0;
                }
                body {
                    background-color: #0F131A;
                }
                ::-webkit-scrollbar {
                    width: 10px;
                }
                ::-webkit-scrollbar-track {
                    background: transparent;
                }
                ::-webkit-scrollbar-thumb {
                    background: #25282e;
                },
                ::-webkit-scrollbar-thumb:hover {
                    background: #393a44;
                }
            `,
        },
    },
})

export default theme
