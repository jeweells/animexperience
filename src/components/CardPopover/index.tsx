import { makeStyles } from '@material-ui/core'
import Portal from '@material-ui/core/Portal'
import { Theme as DefaultTheme } from '@material-ui/core/styles/createMuiTheme'
import React from 'react'
import { Transition } from 'react-transition-group'
import { useSizes } from '../AnimesCarousel/hooks'
import { useContentRef } from '../Topbar'
const useStyles = makeStyles<
    DefaultTheme,
    {
        duration: number
    }
>({
    item: {
        transform: 'scale(1)',
        transition: (props) => `all ${props.duration}ms`,
        '&:hover': {
            transform: 'scale(1.3)',
        },
    },
})
type ScaleOnHoverProps = {
    in: boolean
    children: (className: string) => React.ReactNode
    onExited?(): void
    onEnter?(): void
}

const ScaleOnHover: React.VFC<ScaleOnHoverProps> = ({
    in: inProp,
    children,
    onExited,
    onEnter,
}) => {
    const duration = 300
    const classes = useStyles({
        duration,
    })
    return (
        <Transition
            in={inProp}
            timeout={duration}
            mountOnEnter={true}
            unmountOnExit={true}
            onEnter={onEnter}
            onExited={onExited}
            appear={true}
        >
            {() => {
                return children(classes.item) ?? null
            }}
        </Transition>
    )
}

export type CardPopoverProps = {
    open: boolean
    anchorEl?: React.RefObject<HTMLElement>
    onClose?(): void
}

type Position = {
    top: number
    bottom: number
    left: number
    right: number
}

type MixedPosition = {
    windowRelative: Position
    relative: Position
    containerSize: { width: number; height: number }
}

export const CardPopover: React.FC<CardPopoverProps> = React.memo(
    ({ open, children, anchorEl, onClose }) => {
        const floatingContainerRef = React.useRef<HTMLDivElement>(null)
        const [position, setPosition] = React.useState<MixedPosition | null>(null)
        const { navigationWidth } = useSizes()
        const containerRef = useContentRef()
        const origin = React.useMemo(() => {
            if (position === null) return null
            const { left, right } = position.windowRelative
            const { top, bottom } = position.relative
            let originY: string
            let originX: string
            const wWidth = window.innerWidth
            const threshold = 80
            if (top <= threshold) {
                originY = 'top'
            } else if (position.containerSize.height - bottom <= threshold) {
                originY = 'bottom'
            } else {
                originY = 'center'
            }
            if (left - navigationWidth <= threshold) {
                originX = 'left'
            } else if (wWidth - right - navigationWidth <= threshold) {
                originX = 'right'
            } else {
                originX = 'center'
            }
            return `${originX} ${originY}`
        }, [position, navigationWidth])
        React.useLayoutEffect(() => {
            if (open) {
                const shouldCloseHandle = () => {
                    if (floatingContainerRef.current) {
                        if (!floatingContainerRef.current.matches(':hover')) {
                            // Some cases the popover takes so long to render that
                            // the anchor element is the one that has hover
                            if (!(anchorEl?.current?.matches(':hover') ?? false)) {
                                onClose?.()
                            }
                        }
                    }
                }

                if (anchorEl?.current) {
                    const cb = (cb: DOMRect) => {
                        const containerRect = containerRef?.current?.getBoundingClientRect()
                        if (containerRect) {
                            const ancElRect = cb.toJSON()
                            const scrollTop = containerRef?.current?.scrollTop ?? 0
                            setPosition({
                                // Make it relative to its container
                                relative: {
                                    left: ancElRect.left - containerRect.left,
                                    top: ancElRect.top - containerRect.top + scrollTop,
                                    right: ancElRect.left - containerRect.left,
                                    bottom:
                                        ancElRect.bottom - containerRect.top + scrollTop,
                                },
                                windowRelative: {
                                    left: ancElRect.left,
                                    top: ancElRect.top,
                                    right: ancElRect.right,
                                    bottom: ancElRect.bottom,
                                },
                                containerSize: {
                                    width: containerRect.width,
                                    height: containerRef?.current?.scrollHeight ?? 0,
                                },
                            })
                        }
                    }
                    if (anchorEl.current) {
                        cb(anchorEl.current.getBoundingClientRect())
                    }
                }

                document.addEventListener('mousemove', shouldCloseHandle)
                return () => {
                    document.removeEventListener('mousemove', shouldCloseHandle)
                }
            }
        }, [open])
        return (
            <Portal container={containerRef?.current}>
                <ScaleOnHover
                    in={open}
                    onExited={() => {
                        if (!open) {
                            setPosition(null)
                            onClose?.()
                        }
                    }}
                >
                    {(className) =>
                        position !== null &&
                        origin !== null && (
                            <div
                                ref={floatingContainerRef}
                                style={{
                                    position: 'absolute',
                                    pointerEvents: 'all',
                                    top: position.relative.top,
                                    left: position.relative.left,
                                    zIndex: 100,
                                }}
                            >
                                <div
                                    className={className}
                                    style={{
                                        transformOrigin: origin,
                                    }}
                                >
                                    {children}
                                </div>
                            </div>
                        )
                    }
                </ScaleOnHover>
            </Portal>
        )
    },
)

CardPopover.displayName = 'CardPopover'

export default CardPopover
