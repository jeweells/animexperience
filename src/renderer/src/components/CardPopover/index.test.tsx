import CardPopover, { CardPopoverProps } from './index'
import * as React from 'react'
import { ContentContext } from '@components/Topbar'
import theme from '../../theme'
import { ThemeProvider } from '@mui/material'

describe('CardPopover', () => {
  let props: CardPopoverProps
  const onClose = jest.fn()

  beforeAll(() => {
    jest.spyOn(window.HTMLElement.prototype, 'getBoundingClientRect').mockImplementation(
      () =>
        ({
          bottom: 100,
          height: 100,
          left: 0,
          right: 100,
          top: 0,
          width: 100,
          toJSON() {
            return {
              bottom: 100,
              height: 100,
              left: 0,
              right: 100,
              top: 0,
              width: 100
            }
          }
        }) as DOMRect
    )
  })

  const getComponent = (override: Partial<CardPopoverProps> = {}) => {
    const Wrapper: React.FC = () => {
      const containerRef = React.useRef<HTMLDivElement>(null)
      const anchorRef = React.useRef<HTMLDivElement>(null)
      return (
        <ContentContext.Provider value={containerRef}>
          <div ref={containerRef}>[ContainerRef]</div>
          <div ref={anchorRef}>[AnchorRef]</div>
          <CardPopover {...props} {...override} anchorEl={anchorRef} />
        </ContentContext.Provider>
      )
    }
    Wrapper.displayName = 'WrapperMock'
    return (
      <ThemeProvider theme={theme}>
        <Wrapper />
      </ThemeProvider>
    )
  }

  beforeEach(() => {
    props = {
      open: true,
      onClose,
      children: <div>[Children]</div>
    }
    onClose.mockReset()
  })

  it('renders default', () => {
    const wrapper = render(getComponent())
    expect(wrapper.baseElement).toMatchSnapshot()
  })

  it('renders when closed', () => {
    const wrapper = render(getComponent({ open: false }))
    expect(wrapper.baseElement).toMatchSnapshot()
  })

  it('closes when dropping hover', () => {
    const matchesSpy = jest
      .spyOn(window.Element.prototype, 'matches')
      .mockImplementation(() => false)
    render(getComponent())
    fireEvent.mouseMove(document)
    expect(onClose).toHaveBeenCalledTimes(1)
    matchesSpy.mockRestore()
  })

  it('keeps popover when mouse is hovering', async () => {
    const matchesSpy = jest
      .spyOn(window.Element.prototype, 'matches')
      .mockImplementation(() => true)
    render(getComponent())
    fireEvent.mouseMove(document)
    expect(onClose).toHaveBeenCalledTimes(0)
    matchesSpy.mockRestore()
  })
})
