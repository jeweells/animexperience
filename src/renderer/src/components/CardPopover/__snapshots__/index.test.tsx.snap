// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPopover renders default 1`] = `
<body>
  <div>
    <div>
      [ContainerRef]
      <div
        style="position: absolute; pointer-events: all; top: 0px; left: 0px; z-index: 100;"
      >
        <div
          class="makeStyles-item-1"
          style="transform-origin: left top;"
        >
          <div>
            [Children]
          </div>
        </div>
      </div>
    </div>
    <div>
      [AnchorRef]
    </div>
  </div>
</body>
`;

exports[`CardPopover renders when closed 1`] = `
<body>
  <div>
    <div>
      [ContainerRef]
    </div>
    <div>
      [AnchorRef]
    </div>
  </div>
</body>
`;
